<a href="https://colab.research.google.com/github/Kousikdutta1/Formula-1-Data-Analysis/blob/main/Formula_1_Exploratory_Data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**NAME OF THE DATASET: FORMULA 1 DATASET**

**Formula 1 racing originated during the 1920-30s in Europe from other similar racing competitions. In 1946, the FIA standardized racing rules and this formed the basis of Formula One racing. The inaugural Formula One World Drivers’ championship was then held in 1950, the first world championship series.** 

**Formula 1 also known as F1 is a Motorsport defined by open build single seat race cars it features the best drivers and most powerful and technically advanced cars sanctioned by the Fédération Internationale de l'Automobile also known as FIA. Here 10 team comprising of 2 drivers each compete in a period of one year to find out ultimate driver and a constructor champion. A Formula One season consists of a series of races, known as Grand Prix which take place worldwide on purpose-built circuits and on public roads.**

**The data set includes Race wins, all Race Results, Qualifying Times, Constructors as well as Drivers Championship and Fastest Laps and others for years 1950-2020.**

#**Objective of This Project -**

* Understand the dataset, data types & missing values.
* Clean the dataset and handle the missing values.
* With the help of Data visualization get some conclusions.

#**Importing Libraries**

In [1]:
# We will first import the Python libraries required in this notebook.

import numpy as np  # Helps in working with matrices and arrays
import pandas as pd # Helps in reading the datset , making dataframe and manipulating dataframes
import matplotlib.pyplot as plt  # Helps to visualize the dataset
import seaborn as sns  # Helps to visualize the dataset

%matplotlib inline

#**Uploading the Dataset**

In [2]:
working_dir_path = '/content/drive/MyDrive/Colab Notebooks/Formula 1 Dataset Analysis/'

In [3]:
# Let's load different datasets

results = pd.read_csv(working_dir_path + 'results.csv')

races = pd.read_csv(working_dir_path + 'races.csv')

drivers = pd.read_csv(working_dir_path + 'drivers.csv')

constructors = pd.read_csv(working_dir_path + 'constructors.csv')

constructor_results = pd.read_csv(working_dir_path + 'constructor_results.csv')

circuits = pd.read_csv(working_dir_path + 'circuits.csv')

seasons = pd.read_csv(working_dir_path + 'seasons.csv')

lap_times = pd.read_csv(working_dir_path + 'lap_times.csv')

status = pd.read_csv(working_dir_path + 'status.csv')

##**Let's display top 05 rows of the different datasets of Formula 1 Racing.**

In [4]:
results.head()

resultId  raceId  driverId  constructorId number  grid position  \
0         1      18         1              1     22     1        1   
1         2      18         2              2      3     5        2   
2         3      18         3              3      7     7        3   
3         4      18         4              4      5    11        4   
4         5      18         5              1     23     3        5   

  positionText  positionOrder  points  laps         time milliseconds  \
0            1              1    10.0    58  1:34:50.616      5690616   
1            2              2     8.0    58       +5.478      5696094   
2            3              3     6.0    58       +8.163      5698779   
3            4              4     5.0    58      +17.181      5707797   
4            5              5     4.0    58      +18.014      5708630   

  fastestLap rank fastestLapTime fastestLapSpeed  statusId  
0         39    2       1:27.452         218.300         1  
1         41    3       1:27.739         217.586         1  
2         41    5       1:28.090         216.719         1  
3         58    7       1:28.603         215.464         1  
4         43    1       1:27.418         218.385         1

In [5]:
races.head()

raceId  year  round  circuitId                   name        date  \
0       1  2009      1          1  Australian Grand Prix  2009-03-29   
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05   
2       3  2009      3         17     Chinese Grand Prix  2009-04-19   
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26   
4       5  2009      5          4     Spanish Grand Prix  2009-05-10   

       time                                                url fp1_date  \
0  06:00:00  http://en.wikipedia.org/wiki/2009_Australian_G...       \N   
1  09:00:00  http://en.wikipedia.org/wiki/2009_Malaysian_Gr...       \N   
2  07:00:00  http://en.wikipedia.org/wiki/2009_Chinese_Gran...       \N   
3  12:00:00  http://en.wikipedia.org/wiki/2009_Bahrain_Gran...       \N   
4  12:00:00  http://en.wikipedia.org/wiki/2009_Spanish_Gran...       \N   

  fp1_time fp2_date fp2_time fp3_date fp3_time quali_date quali_time  \
0       \N       \N       \N       \N       \N         \N         \N   
1       \N       \N       \N       \N       \N         \N         \N   
2       \N       \N       \N       \N       \N         \N         \N   
3       \N       \N       \N       \N       \N         \N         \N   
4       \N       \N       \N       \N       \N         \N         \N   

  sprint_date sprint_time  
0          \N          \N  
1          \N          \N  
2          \N          \N  
3          \N          \N  
4          \N          \N

In [6]:
drivers.head()

driverId   driverRef number code  forename     surname         dob  \
0         1    hamilton     44  HAM     Lewis    Hamilton  1985-01-07   
1         2    heidfeld     \N  HEI      Nick    Heidfeld  1977-05-10   
2         3     rosberg      6  ROS      Nico     Rosberg  1985-06-27   
3         4      alonso     14  ALO  Fernando      Alonso  1981-07-29   
4         5  kovalainen     \N  KOV    Heikki  Kovalainen  1981-10-19   

  nationality                                             url  
0     British     http://en.wikipedia.org/wiki/Lewis_Hamilton  
1      German      http://en.wikipedia.org/wiki/Nick_Heidfeld  
2      German       http://en.wikipedia.org/wiki/Nico_Rosberg  
3     Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso  
4     Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen

In [7]:
constructors.head()

constructorId constructorRef        name nationality  \
0              1        mclaren     McLaren     British   
1              2     bmw_sauber  BMW Sauber      German   
2              3       williams    Williams     British   
3              4        renault     Renault      French   
4              5     toro_rosso  Toro Rosso     Italian   

                                                 url  
0               http://en.wikipedia.org/wiki/McLaren  
1            http://en.wikipedia.org/wiki/BMW_Sauber  
2  http://en.wikipedia.org/wiki/Williams_Grand_Pr...  
3  http://en.wikipedia.org/wiki/Renault_in_Formul...  
4   http://en.wikipedia.org/wiki/Scuderia_Toro_Rosso

In [8]:
circuits.head()


circuitId   circuitRef                            name      location  \
0          1  albert_park  Albert Park Grand Prix Circuit     Melbourne   
1          2       sepang    Sepang International Circuit  Kuala Lumpur   
2          3      bahrain   Bahrain International Circuit        Sakhir   
3          4    catalunya  Circuit de Barcelona-Catalunya      Montmeló   
4          5     istanbul                   Istanbul Park      Istanbul   

     country       lat        lng  alt  \
0  Australia -37.84970  144.96800   10   
1   Malaysia   2.76083  101.73800   18   
2    Bahrain  26.03250   50.51060    7   
3      Spain  41.57000    2.26111  109   
4     Turkey  40.95170   29.40500  130   

                                                 url  
0  http://en.wikipedia.org/wiki/Melbourne_Grand_P...  
1  http://en.wikipedia.org/wiki/Sepang_Internatio...  
2  http://en.wikipedia.org/wiki/Bahrain_Internati...  
3  http://en.wikipedia.org/wiki/Circuit_de_Barcel...  
4         http://en.wikipedia.org/wiki/Istanbul_Park

In [9]:
constructor_results.sample(10)

constructorResultsId  raceId  constructorId  points status
4790                   4791     401             32     0.0     \N
11051                 15549     979             15     0.0     \N
9108                  12935     562             67     0.0     \N
557                     558      69             14     0.0     \N
9200                  13212     548             64     0.0     \N
1898                   1899     197             18     0.0     \N
8264                  10009     747            167     0.0     \N
3004                   3005     289             22     3.0     \N
11084                 15582     983              6    12.0     \N
3994                   3995     356             49     0.0     \N

In [10]:
lap_times.head()

raceId  driverId  lap  position      time  milliseconds
0     841        20    1         1  1:38.109         98109
1     841        20    2         1  1:33.006         93006
2     841        20    3         1  1:32.713         92713
3     841        20    4         1  1:32.803         92803
4     841        20    5         1  1:32.342         92342

In [11]:
status.head()

statusId        status
0         1      Finished
1         2  Disqualified
2         3      Accident
3         4     Collision
4         5        Engine

In [12]:
seasons.head()

year                                                url
0  2009  http://en.wikipedia.org/wiki/2009_Formula_One_...
1  2008  http://en.wikipedia.org/wiki/2008_Formula_One_...
2  2007  http://en.wikipedia.org/wiki/2007_Formula_One_...
3  2006  http://en.wikipedia.org/wiki/2006_Formula_One_...
4  2005  http://en.wikipedia.org/wiki/2005_Formula_One_...

#**Let's merge different datasets for better understanding.**

In [13]:
# Let's first merge results and races column.

merged_results_races = pd.merge (results, races, on = 'raceId', how = 'left')

In [14]:
# Let's view what are the columns that are present in the merged dataset.

merged_results_races.columns

Index(['resultId', 'raceId', 'driverId', 'constructorId', 'number', 'grid',
       'position', 'positionText', 'positionOrder', 'points', 'laps', 'time_x',
       'milliseconds', 'fastestLap', 'rank', 'fastestLapTime',
       'fastestLapSpeed', 'statusId', 'year', 'round', 'circuitId', 'name',
       'date', 'time_y', 'url', 'fp1_date', 'fp1_time', 'fp2_date', 'fp2_time',
       'fp3_date', 'fp3_time', 'quali_date', 'quali_time', 'sprint_date',
       'sprint_time'],
      dtype='object')

In [15]:
# Let's view top 5 row of the dataset.

merged_results_races.head()

resultId  raceId  driverId  constructorId number  grid position  \
0         1      18         1              1     22     1        1   
1         2      18         2              2      3     5        2   
2         3      18         3              3      7     7        3   
3         4      18         4              4      5    11        4   
4         5      18         5              1     23     3        5   

  positionText  positionOrder  points  ...  fp1_date fp1_time fp2_date  \
0            1              1    10.0  ...        \N       \N       \N   
1            2              2     8.0  ...        \N       \N       \N   
2            3              3     6.0  ...        \N       \N       \N   
3            4              4     5.0  ...        \N       \N       \N   
4            5              5     4.0  ...        \N       \N       \N   

  fp2_time fp3_date fp3_time quali_date  quali_time  sprint_date  sprint_time  
0       \N       \N       \N         \N          \N           \N           \N  
1       \N       \N       \N         \N          \N           \N           \N  
2       \N       \N       \N         \N          \N           \N           \N  
3       \N       \N       \N         \N          \N           \N           \N  
4       \N       \N       \N         \N          \N           \N           \N  

[5 rows x 35 columns]

In [16]:
# Also we can merge driver dataset with merged_results_races dataset.

merged_results_races_driver = pd.merge (merged_results_races,drivers, on = 'driverId', how = 'left')

In [17]:
# Let's view the merged dataset.

merged_results_races_driver.head()

resultId  raceId  driverId  constructorId number_x  grid position  \
0         1      18         1              1       22     1        1   
1         2      18         2              2        3     5        2   
2         3      18         3              3        7     7        3   
3         4      18         4              4        5    11        4   
4         5      18         5              1       23     3        5   

  positionText  positionOrder  points  ...  sprint_date sprint_time  \
0            1              1    10.0  ...           \N          \N   
1            2              2     8.0  ...           \N          \N   
2            3              3     6.0  ...           \N          \N   
3            4              4     5.0  ...           \N          \N   
4            5              5     4.0  ...           \N          \N   

    driverRef number_y code  forename     surname         dob  nationality  \
0    hamilton       44  HAM     Lewis    Hamilton  1985-01-07      British   
1    heidfeld       \N  HEI      Nick    Heidfeld  1977-05-10       German   
2     rosberg        6  ROS      Nico     Rosberg  1985-06-27       German   
3      alonso       14  ALO  Fernando      Alonso  1981-07-29      Spanish   
4  kovalainen       \N  KOV    Heikki  Kovalainen  1981-10-19      Finnish   

                                            url_y  
0     http://en.wikipedia.org/wiki/Lewis_Hamilton  
1      http://en.wikipedia.org/wiki/Nick_Heidfeld  
2       http://en.wikipedia.org/wiki/Nico_Rosberg  
3    http://en.wikipedia.org/wiki/Fernando_Alonso  
4  http://en.wikipedia.org/wiki/Heikki_Kovalainen  

[5 rows x 43 columns]

In [18]:
# Let's view what are the columns that are present in the merged dataset.

merged_results_races_driver.columns

Index(['resultId', 'raceId', 'driverId', 'constructorId', 'number_x', 'grid',
       'position', 'positionText', 'positionOrder', 'points', 'laps', 'time_x',
       'milliseconds', 'fastestLap', 'rank', 'fastestLapTime',
       'fastestLapSpeed', 'statusId', 'year', 'round', 'circuitId', 'name',
       'date', 'time_y', 'url_x', 'fp1_date', 'fp1_time', 'fp2_date',
       'fp2_time', 'fp3_date', 'fp3_time', 'quali_date', 'quali_time',
       'sprint_date', 'sprint_time', 'driverRef', 'number_y', 'code',
       'forename', 'surname', 'dob', 'nationality', 'url_y'],
      dtype='object')

In [19]:
# Again merge the merged_results_races_driver dataset with the constructor dataset.

merged_results_races_driver_constructor = pd.merge (merged_results_races_driver, constructors, on = 'constructorId', how = 'left')

In [20]:
# Let's view the merged dataset.

merged_results_races_driver_constructor.head()

resultId  raceId  driverId  constructorId number_x  grid position  \
0         1      18         1              1       22     1        1   
1         2      18         2              2        3     5        2   
2         3      18         3              3        7     7        3   
3         4      18         4              4        5    11        4   
4         5      18         5              1       23     3        5   

  positionText  positionOrder  points  ...  code  forename     surname  \
0            1              1    10.0  ...   HAM     Lewis    Hamilton   
1            2              2     8.0  ...   HEI      Nick    Heidfeld   
2            3              3     6.0  ...   ROS      Nico     Rosberg   
3            4              4     5.0  ...   ALO  Fernando      Alonso   
4            5              5     4.0  ...   KOV    Heikki  Kovalainen   

          dob nationality_x                                           url_y  \
0  1985-01-07       British     http://en.wikipedia.org/wiki/Lewis_Hamilton   
1  1977-05-10        German      http://en.wikipedia.org/wiki/Nick_Heidfeld   
2  1985-06-27        German       http://en.wikipedia.org/wiki/Nico_Rosberg   
3  1981-07-29       Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso   
4  1981-10-19       Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen   

  constructorRef      name_y  nationality_y  \
0        mclaren     McLaren        British   
1     bmw_sauber  BMW Sauber         German   
2       williams    Williams        British   
3        renault     Renault         French   
4        mclaren     McLaren        British   

                                                 url  
0               http://en.wikipedia.org/wiki/McLaren  
1            http://en.wikipedia.org/wiki/BMW_Sauber  
2  http://en.wikipedia.org/wiki/Williams_Grand_Pr...  
3  http://en.wikipedia.org/wiki/Renault_in_Formul...  
4               http://en.wikipedia.org/wiki/McLaren  

[5 rows x 47 columns]

In [21]:
# Let's view what are the columns that are present in the merged dataset.


merged_results_races_driver_constructor.columns

Index(['resultId', 'raceId', 'driverId', 'constructorId', 'number_x', 'grid',
       'position', 'positionText', 'positionOrder', 'points', 'laps', 'time_x',
       'milliseconds', 'fastestLap', 'rank', 'fastestLapTime',
       'fastestLapSpeed', 'statusId', 'year', 'round', 'circuitId', 'name_x',
       'date', 'time_y', 'url_x', 'fp1_date', 'fp1_time', 'fp2_date',
       'fp2_time', 'fp3_date', 'fp3_time', 'quali_date', 'quali_time',
       'sprint_date', 'sprint_time', 'driverRef', 'number_y', 'code',
       'forename', 'surname', 'dob', 'nationality_x', 'url_y',
       'constructorRef', 'name_y', 'nationality_y', 'url'],
      dtype='object')

In [22]:
# Finally merge the merged_results_races_driver_constructor dataset with the status dataset.

merged_results_races_driver_constructor_status = pd.merge (merged_results_races_driver_constructor, status, on = 'statusId', how = 'left')

In [23]:
# Let's view the merged dataset.

merged_results_races_driver_constructor_status.head()

resultId  raceId  driverId  constructorId number_x  grid position  \
0         1      18         1              1       22     1        1   
1         2      18         2              2        3     5        2   
2         3      18         3              3        7     7        3   
3         4      18         4              4        5    11        4   
4         5      18         5              1       23     3        5   

  positionText  positionOrder  points  ...  forename     surname         dob  \
0            1              1    10.0  ...     Lewis    Hamilton  1985-01-07   
1            2              2     8.0  ...      Nick    Heidfeld  1977-05-10   
2            3              3     6.0  ...      Nico     Rosberg  1985-06-27   
3            4              4     5.0  ...  Fernando      Alonso  1981-07-29   
4            5              5     4.0  ...    Heikki  Kovalainen  1981-10-19   

  nationality_x                                           url_y  \
0       British     http://en.wikipedia.org/wiki/Lewis_Hamilton   
1        German      http://en.wikipedia.org/wiki/Nick_Heidfeld   
2        German       http://en.wikipedia.org/wiki/Nico_Rosberg   
3       Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso   
4       Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen   

  constructorRef      name_y  nationality_y  \
0        mclaren     McLaren        British   
1     bmw_sauber  BMW Sauber         German   
2       williams    Williams        British   
3        renault     Renault         French   
4        mclaren     McLaren        British   

                                                 url    status  
0               http://en.wikipedia.org/wiki/McLaren  Finished  
1            http://en.wikipedia.org/wiki/BMW_Sauber  Finished  
2  http://en.wikipedia.org/wiki/Williams_Grand_Pr...  Finished  
3  http://en.wikipedia.org/wiki/Renault_in_Formul...  Finished  
4               http://en.wikipedia.org/wiki/McLaren  Finished  

[5 rows x 48 columns]

In [24]:
# Let's find the shape of the final dataset.

merged_results_races_driver_constructor_status.shape

(25620, 48)

In [25]:
# Let's find the columns of the final dataset and identify those columns which are not needed.



merged_results_races_driver_constructor_status.columns


Index(['resultId', 'raceId', 'driverId', 'constructorId', 'number_x', 'grid',
       'position', 'positionText', 'positionOrder', 'points', 'laps', 'time_x',
       'milliseconds', 'fastestLap', 'rank', 'fastestLapTime',
       'fastestLapSpeed', 'statusId', 'year', 'round', 'circuitId', 'name_x',
       'date', 'time_y', 'url_x', 'fp1_date', 'fp1_time', 'fp2_date',
       'fp2_time', 'fp3_date', 'fp3_time', 'quali_date', 'quali_time',
       'sprint_date', 'sprint_time', 'driverRef', 'number_y', 'code',
       'forename', 'surname', 'dob', 'nationality_x', 'url_y',
       'constructorRef', 'name_y', 'nationality_y', 'url', 'status'],
      dtype='object')

In [26]:
# We can make a copy of the original dataset on which we will be doing all analysis.

f1df = merged_results_races_driver_constructor_status.copy()

In [27]:
# Let's visualize all the row of the dataset

pd.set_option('display.max_columns', None)
f1df.head()

resultId  raceId  driverId  constructorId number_x  grid position  \
0         1      18         1              1       22     1        1   
1         2      18         2              2        3     5        2   
2         3      18         3              3        7     7        3   
3         4      18         4              4        5    11        4   
4         5      18         5              1       23     3        5   

  positionText  positionOrder  points  laps       time_x milliseconds  \
0            1              1    10.0    58  1:34:50.616      5690616   
1            2              2     8.0    58       +5.478      5696094   
2            3              3     6.0    58       +8.163      5698779   
3            4              4     5.0    58      +17.181      5707797   
4            5              5     4.0    58      +18.014      5708630   

  fastestLap rank fastestLapTime fastestLapSpeed  statusId  year  round  \
0         39    2       1:27.452         218.300         1  2008      1   
1         41    3       1:27.739         217.586         1  2008      1   
2         41    5       1:28.090         216.719         1  2008      1   
3         58    7       1:28.603         215.464         1  2008      1   
4         43    1       1:27.418         218.385         1  2008      1   

   circuitId                 name_x        date    time_y  \
0          1  Australian Grand Prix  2008-03-16  04:30:00   
1          1  Australian Grand Prix  2008-03-16  04:30:00   
2          1  Australian Grand Prix  2008-03-16  04:30:00   
3          1  Australian Grand Prix  2008-03-16  04:30:00   
4          1  Australian Grand Prix  2008-03-16  04:30:00   

                                               url_x fp1_date fp1_time  \
0  http://en.wikipedia.org/wiki/2008_Australian_G...       \N       \N   
1  http://en.wikipedia.org/wiki/2008_Australian_G...       \N       \N   
2  http://en.wikipedia.org/wiki/2008_Australian_G...       \N       \N   
3  http://en.wikipedia.org/wiki/2008_Australian_G...       \N       \N   
4  http://en.wikipedia.org/wiki/2008_Australian_G...       \N       \N   

  fp2_date fp2_time fp3_date fp3_time quali_date quali_time sprint_date  \
0       \N       \N       \N       \N         \N         \N          \N   
1       \N       \N       \N       \N         \N         \N          \N   
2       \N       \N       \N       \N         \N         \N          \N   
3       \N       \N       \N       \N         \N         \N          \N   
4       \N       \N       \N       \N         \N         \N          \N   

  sprint_time   driverRef number_y code  forename     surname         dob  \
0          \N    hamilton       44  HAM     Lewis    Hamilton  1985-01-07   
1          \N    heidfeld       \N  HEI      Nick    Heidfeld  1977-05-10   
2          \N     rosberg        6  ROS      Nico     Rosberg  1985-06-27   
3          \N      alonso       14  ALO  Fernando      Alonso  1981-07-29   
4          \N  kovalainen       \N  KOV    Heikki  Kovalainen  1981-10-19   

  nationality_x                                           url_y  \
0       British     http://en.wikipedia.org/wiki/Lewis_Hamilton   
1        German      http://en.wikipedia.org/wiki/Nick_Heidfeld   
2        German       http://en.wikipedia.org/wiki/Nico_Rosberg   
3       Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso   
4       Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen   

  constructorRef      name_y nationality_y  \
0        mclaren     McLaren       British   
1     bmw_sauber  BMW Sauber        German   
2       williams    Williams       British   
3        renault     Renault        French   
4        mclaren     McLaren       British   

                                                 url    status  
0               http://en.wikipedia.org/wiki/McLaren  Finished  
1            http://en.wikipedia.org/wiki/BMW_Sauber  Finished  
2  http://en.wikipedia.org/wiki/Williams_Grand_Pr...  Finished  
3  http://en.w

In [28]:
# Let's check all the information about the dataset

f1df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25620 entries, 0 to 25619
Data columns (total 48 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         25620 non-null  int64  
 1   raceId           25620 non-null  int64  
 2   driverId         25620 non-null  int64  
 3   constructorId    25620 non-null  int64  
 4   number_x         25620 non-null  object 
 5   grid             25620 non-null  int64  
 6   position         25620 non-null  object 
 7   positionText     25620 non-null  object 
 8   positionOrder    25620 non-null  int64  
 9   points           25620 non-null  float64
 10  laps             25620 non-null  int64  
 11  time_x           25620 non-null  object 
 12  milliseconds     25620 non-null  object 
 13  fastestLap       25620 non-null  object 
 14  rank             25620 non-null  object 
 15  fastestLapTime   25620 non-null  object 
 16  fastestLapSpeed  25620 non-null  object 
 17  statusId    

* **As we can see that there are no null values in the dataset, all 47 columns each have 25620 non-null values present.**

In [29]:
# checking of null values

f1df.isna().sum()

resultId           0
raceId             0
driverId           0
constructorId      0
number_x           0
grid               0
position           0
positionText       0
positionOrder      0
points             0
laps               0
time_x             0
milliseconds       0
fastestLap         0
rank               0
fastestLapTime     0
fastestLapSpeed    0
statusId           0
year               0
round              0
circuitId          0
name_x             0
date               0
time_y             0
url_x              0
fp1_date           0
fp1_time           0
fp2_date           0
fp2_time           0
fp3_date           0
fp3_time           0
quali_date         0
quali_time         0
sprint_date        0
sprint_time        0
driverRef          0
number_y           0
code               0
forename           0
surname            0
dob                0
nationality_x      0
url_y              0
constructorRef     0
name_y             0
nationality_y      0
url                0
status       

* **We can see from above all the columns have 0 null values.** 

In [30]:
# Let's check all the statistical description about the dataset.
# Describe function is used to show the statistical value of each column.

f1df.describe()

resultId        raceId      driverId  constructorId          grid  \
count  25620.000000  25620.000000  25620.000000   25620.000000  25620.000000   
mean   12811.298556    526.624629    257.847268      48.225839     11.189891   
std     7397.131226    296.187675    265.045568      59.269433      7.253758   
min        1.000000      1.000000      1.000000       1.000000      0.000000   
25%     6405.750000    291.000000     56.000000       6.000000      5.000000   
50%    12810.500000    510.000000    160.000000      25.000000     11.000000   
75%    19215.250000    776.000000    357.000000      58.000000     17.000000   
max    25625.000000   1084.000000    855.000000     214.000000     34.000000   

       positionOrder        points          laps      statusId          year  \
count   25620.000000  25620.000000  25620.000000  25620.000000  25620.000000   
mean       12.896409      1.849456     45.916784     17.612178   1989.979157   
std         7.723816      4.122851     29.882464     26.090977     19.202536   
min         1.000000      0.000000      0.000000      1.000000   1950.000000   
25%         6.000000      0.000000     22.000000      1.000000   1976.000000   
50%        12.000000      0.000000     52.000000     11.000000   1990.000000   
75%        18.000000      2.000000     66.000000     14.000000   2006.000000   
max        39.000000     50.000000    200.000000    140.000000   2022.000000   

              round     circuitId  
count  25620.000000  25620.000000  
mean       8.312022     23.461436  
std        4.905754     18.586948  
min        1.000000      1.000000  
25%        4.000000      9.000000  
50%        8.000000     18.000000  
75%       12.000000     34.000000  
max       22.000000     79.000000

In [31]:
# Let's drop the unnecessary columns.


f1df.drop(['position', 'positionText', 'time_x', 'time_y', 'url_x', 'fp1_date', 'fp1_time', 'fp2_date', 'fp2_time', 'fp3_date', 'fp3_time', 'quali_date', 'quali_time',
           'sprint_date', 'sprint_time', 'url_y', 'url'], inplace = True, axis = 1)


In [32]:
# Let's check the shape of the dataset

f1df.shape

(25620, 31)

In [33]:
# Let's check the top rows of the dataset

f1df.head()

resultId  raceId  driverId  constructorId number_x  grid  positionOrder  \
0         1      18         1              1       22     1              1   
1         2      18         2              2        3     5              2   
2         3      18         3              3        7     7              3   
3         4      18         4              4        5    11              4   
4         5      18         5              1       23     3              5   

   points  laps milliseconds fastestLap rank fastestLapTime fastestLapSpeed  \
0    10.0    58      5690616         39    2       1:27.452         218.300   
1     8.0    58      5696094         41    3       1:27.739         217.586   
2     6.0    58      5698779         41    5       1:28.090         216.719   
3     5.0    58      5707797         58    7       1:28.603         215.464   
4     4.0    58      5708630         43    1       1:27.418         218.385   

   statusId  year  round  circuitId                 name_x        date  \
0         1  2008      1          1  Australian Grand Prix  2008-03-16   
1         1  2008      1          1  Australian Grand Prix  2008-03-16   
2         1  2008      1          1  Australian Grand Prix  2008-03-16   
3         1  2008      1          1  Australian Grand Prix  2008-03-16   
4         1  2008      1          1  Australian Grand Prix  2008-03-16   

    driverRef number_y code  forename     surname         dob nationality_x  \
0    hamilton       44  HAM     Lewis    Hamilton  1985-01-07       British   
1    heidfeld       \N  HEI      Nick    Heidfeld  1977-05-10        German   
2     rosberg        6  ROS      Nico     Rosberg  1985-06-27        German   
3      alonso       14  ALO  Fernando      Alonso  1981-07-29       Spanish   
4  kovalainen       \N  KOV    Heikki  Kovalainen  1981-10-19       Finnish   

  constructorRef      name_y nationality_y    status  
0        mclaren     McLaren       British  Finished  
1     bmw_sauber  BMW Sauber        German  Finished  
2       williams    Williams       British  Finished  
3        renault     Renault        French  Finished  
4        mclaren     McLaren       British  Finished

In [34]:
# Now we can also rename the dataset for the better understanding

f1df.rename(columns = {'number_x':'number','milliseconds':'timetaken_inms','fastestLapSpeed':'maximum_speed',
 'name_x':'grand_prix','number_y':'driver_number','code':'driver_code','nationality_x':'nationality','name_y':'company', 'rank': 'fastest_lap_rank'
 }, inplace = True)

In [35]:
# Let's check the column names of the dataset

f1df.columns

Index(['resultId', 'raceId', 'driverId', 'constructorId', 'number', 'grid',
       'positionOrder', 'points', 'laps', 'timetaken_inms', 'fastestLap',
       'fastest_lap_rank', 'fastestLapTime', 'maximum_speed', 'statusId',
       'year', 'round', 'circuitId', 'grand_prix', 'date', 'driverRef',
       'driver_number', 'driver_code', 'forename', 'surname', 'dob',
       'nationality', 'constructorRef', 'company', 'nationality_y', 'status'],
      dtype='object')

In [36]:
f1df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25620 entries, 0 to 25619
Data columns (total 31 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   resultId          25620 non-null  int64  
 1   raceId            25620 non-null  int64  
 2   driverId          25620 non-null  int64  
 3   constructorId     25620 non-null  int64  
 4   number            25620 non-null  object 
 5   grid              25620 non-null  int64  
 6   positionOrder     25620 non-null  int64  
 7   points            25620 non-null  float64
 8   laps              25620 non-null  int64  
 9   timetaken_inms    25620 non-null  object 
 10  fastestLap        25620 non-null  object 
 11  fastest_lap_rank  25620 non-null  object 
 12  fastestLapTime    25620 non-null  object 
 13  maximum_speed     25620 non-null  object 
 14  statusId          25620 non-null  int64  
 15  year              25620 non-null  int64  
 16  round             25620 non-null  int64 

In [37]:
# Let's replace the 'n' values with np.nan

f1df['number'].replace('\\N', np.nan, inplace = True)

In [38]:
# Let's change the data type of number column to float

f1df['number'] = f1df['number'].astype(float)

In [39]:


f1df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25620 entries, 0 to 25619
Data columns (total 31 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   resultId          25620 non-null  int64  
 1   raceId            25620 non-null  int64  
 2   driverId          25620 non-null  int64  
 3   constructorId     25620 non-null  int64  
 4   number            25614 non-null  float64
 5   grid              25620 non-null  int64  
 6   positionOrder     25620 non-null  int64  
 7   points            25620 non-null  float64
 8   laps              25620 non-null  int64  
 9   timetaken_inms    25620 non-null  object 
 10  fastestLap        25620 non-null  object 
 11  fastest_lap_rank  25620 non-null  object 
 12  fastestLapTime    25620 non-null  object 
 13  maximum_speed     25620 non-null  object 
 14  statusId          25620 non-null  int64  
 15  year              25620 non-null  int64  
 16  round             25620 non-null  int64 